Updating changes to guide for review (#5732)
aws-amplify/docs: src/fragments/lib/graphqlapi/js/setup/connect.mdx

> > *This guide provides essential information to help you select and complete the activities you need for your Amplify project. We organized each section around a distinct job or decision point to help you understand your available options, steps to complete, and recommended best practices. For users who want to learn more, we included extra details and explanations in expandable sections within the guide.*

Author: General notes here -

This guide is BIG even with the Expanders. We may want to revisit breaking the jobs up and adding a thread/landing page that we talked about before to break this up more or moving some of this content to future guides and slimming down some of the guidance here.
The updated expander is still WIP so keep in mind that:
-- the menu on the right is not correct and expander options will not show there in the future
-- Expanders will have a close footer available if users want to close but not scroll up
-- Expander headings will look the same regardless if H3 or H4
-- Animation for closing the Expand is also being fixed (won't see the text issues)

In this guide, you will learn how to connect an existing data stack to your application. This job includes understanding the required client configuration and how to connect the data stack. We will also review functions to create, read, update, and delete your data using the Amplify GraphQL client. Finally, we will look at how to subscribe to real-time events to watch for mutations in your data.

> Before you begin, you will need:

Author: Note - When reviewing the Auth guide we determined it made more sense to have prereqs at the top as it was getting confusing list and repeating them for each section. So I moved this up and other prereqs now reference the section that needs to be completed rather than repeating steps. Updated this guide with that change so we can keep the template consistent - if it does not work we are opening to revisiting!

| **Recommended use case** | **Strategy** | **Provider** |
|---|---|---|
| Public data access where users or devices are anonymous. Anyone with the AWS AppSync API key is granted access. | [`public`]() | `apiKey` |
| Recommended for public data access in a production environment, or where unauthenticated users or devices are granted permissions using AWS IAM controls. | [`public`]() | `iam` |
| Per-user data access. Access is restricted to the "owner" of a record. Leverages `amplify add auth` Cognito user pool by default. | [`owner`]() | `userPools` / `oidc` |
| Any signed-in data access. Unlike owner-based access, **any** signed-in user has access. | [`private`]() | `userPools` / `oidc` / `iam` |
| Per-user group data access. A specific or dynamically configured group of users have access. | [`group`]() | `userPools` / `oidc` |
| Define your own custom authorization rule within a Lambda function. | [`custom`]() | `function` |

</Accordion>

<br/>

Choose your authentication type from the tabs below and configure the client by updating the following details:

<BlockSwitcher>
<Block name="API_KEY">

> 

Author: Q from Jim during his initial edit pass - We sometimes use javascript and other times use js at the beginning of code blocks. Do we need to be consistent? Same question applies to the TypeScript blocks.

Reviewer: no need to be consistent. It's whatever's convenient. In the end the tool will automatically map js -> javascript and ts -> typescript.

Amplify.configure(myAppConfig);
```

> 

Author: Q from Jim during his initial edit pass - In line 81 above, should "ID" be all caps in both instances?

Reviewer: Yeah we should keep it consistent.

Amplify.configure(myAppConfig);
```

> 

Author: Q from Jim during his initial edit pass - I realize this is likely outside the scope of this review, but the message in line 97 above within the code block - Before calling API,graphql... - isn't particularly clear. Is this something we can update to make clearer for the user?

</Block>
</BlockSwitcher>

#### Using with an AWS AppSync custom domain name

> [Custom domain names](https://docs.aws.amazon.com/appsync/latest/devguide/custom-domain-name.html) can be used to create a memorable endpoint with a domain name of your choice. These can have any format, but must end with `/graphql` (see [URIs, Routes](https://graphql.org/learn/serving-over-http/#uris-routes) on the GraphQL website). Using custom domain names will impact how you enter your `graphqlEndpoint`:

Author: Q from Jim during his initial edit pass - In the previous line, can we be more specific about how use of the custom domain name affects how you enter your graphqlEndpoint? How is this demonstrated in the following code?

Reviewer: We could update the Line 114 with the following comment

```
  aws_appsync_graphqlEndpoint: 'https://api.yourdomain.com/graphql', // <- Provide your custom domain here
```

Amplify.configure(myAppConfig);
```

> <Accordion title='Using a non-AppSync GraphQL server' headingLevel='3' eyebrow='Learn more'>

Author: I missed asking this Q again in our Slack conversation but I made the choice to put this next section in an expander vs on the page. Does this work or do we want to take this option out of the expander and list AppSync as Option 1 and non-AppSync as Option 2?

Reviewer: Okay as expander. It's not a recommended path for most customers.

});
```

</Accordion>

<br/>

> **Recap:** At this point, your client is now configured to point to your AWS AppSync API endpoint and you updated your authentication details based on the auth types you are using.

Author: Is there anything more we want to add to the Recap to illustrate a tie-in to real-world use or application?

Reviewer: Maybe we can say also "Now you're ready to call your GraphQL API." Something that alludes to that this is a pre-requisite to get things set up.

## Connect your application code to the data backend

You will connect your application code to the data stack by connecting and configuring the Amplify Libraries. We will review how to install and configure Amplify Libraries to help you complete this task. Before starting these steps you should [review the required configuration](#Review-the-required-configuration).

> 

Author: This section below reads out of place and needs a better tie in to the intro above and what follows. From the intro I was expecting to dive right into connecting and configuring libraries not using codegen. Can we update this section to work together more so it reads less out of place? Also, is this only for JS/TS or do we have it for all frameworks? Just wondering if this section will have different content for different frameworks?

Reviewer: I agree this reads out of place and I think the placement of this might change as I wrap up the other sections. I'd recommend adding a "TODO" to come back later to see the best place to put this.

> 

Author: Same here with this sentence abive - can we expand on this - why am I doing this step here vs later with the next section? Can add detail on why it is needed in this order?

</Block>

<Block name="NextJS">

The first step to using Amplify in the client for NextJS is to install the necessary dependencies:

```bash
npm install aws-amplify @aws-amplify/ui-react
```

The `aws-amplify` package is the main library for working with Amplify in your apps. The `@aws-amplify/ui-react` package includes React-specific UI components that you can use to build your app UI. To learn more about Amplify UI, visit our [documentation](https://ui.docs.amplify.aws/).

> 

Author: Q on the UI link here - is there a better link we can give them - is there a specific part of the docs they should be looking at?

</Block>
</BlockSwitcher>

> **Recap:** Well done! You now have your Amplify Libraries set up and configured. 

Author: Is there anything more we want to add to the Recap above to illustrate a tie-in to real-world use or application?

## Create, update, and delete application data

In this job, you will run mutations to create, update, and delete application data. We will also show you how to cancel these requests. Before you begin, you will need your application connected to the API.

### Run mutations to create, update, and delete application data
  

> In GraphQL, mutations write data to the API and are used to create, update, or delete data. This is different than queries that allow you to read the data but not change it. What follows are some examples of how you can create, update, and delete items using the Amplify GraphQL client.

Author: Note here that I thought an expander seemed like too much to remind them of the diff b/w mutations and queries. I added the brief review directly into this intro. We can change it back to something more robust if needed.

Reviewer: Acknowledged. This works.

</BlockSwitcher>

> <Callout>

Author: Put this one in a note callout to keep consistent with one that follows below.

Reviewer: sounds good.

</BlockSwitcher>

<Callout>

__Notes:__

- You do not have to pass in the `updatedAt` field. AWS AppSync manages this for you.
- If you pass in *extra* input fields not expected by the AWS AppSync schema, this query will fail. You can see this in the `error` field returned by the query. In GraphQL, errors are not thrown like exceptions in other languages. Instead, any errors are captured and returned as part of the query result in the `error` field.

</Callout>

#### Delete an item

> You can then delete the Todo by using the delete mutation. Only an `id` is needed to identify which item to delete:

Author: I moved up the sentence after the code block about the id b/c it seemed out of place at the bottom - does it need to move back?

</BlockSwitcher>

<Callout>

__Note:__ Join table records must be deleted prior to deleting the associated records. For example, for a many-to-many relationship between Posts and Tags, delete the PostTags join record prior to deleting a Post or Tag.

</Callout>

> <Accordion title='Custom authorization mode to mutate data' headingLevel='4' eyebrow='Learn more'>

Author: Kept the Learn more eyebrow text instead of troubleshooting on the expander below b/c this reads as more of an option. We originally had a note to provide troubleshooting for unauthorized errors and cross-platform gotchas - this does not read to me like it covers those errors. Additionally I think the intro within this expander could add an explanation of why a dev may want to override this

</Block>
</BlockSwitcher>

</Accordion>

### Cancel mutation requests

> You may cancel any query or mutation request made through the API category of Amplify by keeping a reference to the promise returned.

Author: Q from Jim during his initial edit pass - Will this sentence be clear to all readers? Can we spell this out in more detail?

Reviewer: ```suggestion
You can cancel any GraphQL API request calling `.cancel` on the GraphQL request promise that's returned by `API.graphql(...)`.
```

API.cancel(promise, "my error message");
```

> **Recap:** You have implemented a few mutations to change your data. 

Author: More detail needed here? We're not really addressing the tie-in or real-world use case or the benefit of what we've done in this module.

Reviewer: You are now able to create, update, and delete your app data via the GraphQL API.

## Read application data

> You can read application data using the Amplify GraphQL client. In this next section we will review the difference between listing data and getting data, how to filter query results to get just the data you need, and how to paginate results to make you're data more manageable. We will also show you how to cancel these requests when needed. Before you begin you will need your application connected to the API and some populated data to query.

Author: Again we are changing the template to have the starting prereqs at the top and then reference the other sections. I did not link this one and the previous since these will be modularized. But users can find this information on the same page. I also removed the proposed bullet points to keep the narrative intro structure.

Reviewer: LGTM.

### Read and filter your data
  
Queries are used to read data from the API and include the `List` and `Get` operations.
  
You can use both `List` and `Get` to query data in GraphQL. A `List` will get you a record of several items, such as Todo items, and does not require an ID. This is best suited for getting an overview or summary of items or can help when filtering by specific criteria across a list of items. When you want to query a single entry by ID, you would use `Get` to retrieve a specific Todo item.

> <Callout>

Author: Changed this to a note vs an Expander since a small section.

__Note:__ The cost structure of your underlying data source can impact the cost to run some queries. For example, the `List` operation uses "scan operations" which uses more read request units than the `Get` operation. You will want to review the associated costs for these operations for your data source. In our example case here we are using Amazon DynamoDB and you can [learn more about how DynamoDB pricing is calculated](https://aws.amazon.com/dynamodb/pricing/) on their site.

</Callout>

> The Amplify CLI's codegen automatically creates code for all possible GraphQL operations—mutations, queries, and subscriptions. For JavaScript applications, this generated code is saved in the src/graphql folder. 

Author: Again general question about Codegen is it only available for JS/TS - should we be considering this section is replaced for other frameworks?

Reviewer: it's available for all frameworks but I think we need to revisit where to "place" this content in general.

##### Compound filters

You can combine filters with `and`, `or`, and `not` Boolean logic. Observe, in the autogenerated schema above, that `ModelTodoFilterInput` is recursive in respect to those fields. So if, for example, you wanted to filter for `priority` values of 1 OR 2, you would do this:

> [comment]: # (HEATHER, this paragraph references an earlier section. We'll need to discuss how we want to handle this in terms of the modularization approach.)

Author: Just noting there was a Q about reference to earlier section and impact on modularization - this all looks to be part of the same H2 so we should be good

#### Paginating queries

> To paginate results in AWS AppSync, make a query request specifying a `limit` value. This limits how many results are returned. The response will include a `nextToken` you can use to request the next page of data. A `nextToken` is a very long string that represents the cursor to the starting item of the next query made with these filters.

Author: Note - used Jim's suggestion for rewrite and moved up as the description made more sense before the example code.

</Block>
</BlockSwitcher>

> <Accordion title='Implement pagination in your GraphQL API' headingLevel='3' eyebrow='Walkthrough'>

Author: Moved the guide info here to an expander since we noted it is still relevant and one of the goals was to eventually move all guide content to libraries. However, we may want to review where this may also live on other JTBD Guides.

  query listTodos($limit: Int) {
    listTodos(limit: $limit) {
      items {
        id
        title
        description
      }
      nextToken
    }
  }
`
```

To pass in a `limit` in a query from a JavaScript application, you can use the following code by setting the limit as a variable:

```js
import { API } from 'aws-amplify';

async function fetchTodos() {
  const todoData = await API.graphql({
    query: listTodos,
    variables: {
      limit: 2
    }
  })
  console.log({ todoData })
}
```

The data returned from the API request should look like this (with the items array containing however many items have been created):

```graphql
{
  "data" {
    "listTodos" {
      "items": [{ id: "001", title: "Todo 1", description: "My first todo" }],
      "nextToken": "<token-id>"
    }
  }
}
```

To set the `nextToken` in a query from a JavaScript application, you can use the following code:

```js
import { API } from 'aws-amplify';

async function fetchTodos() {
  const todoData = await API.graphql({
    query: listTodos,
    variables: {
      limit: 2,
      nextToken: "<token-id>"
    }
  })
  console.log({ todoData })
}
```

</Accordion>

<br/>

<Callout>

> __Limitations:__

Author: Included the information about nextToken discussed in Slack and other limitations all together here.

- There is no API to get a total page count at this time. Note that scanning all items is a [potentially expensive operation](https://github.com/aws-amplify/amplify-js/issues/2901).
- Sorting is [available in DataStore](https://docs.amplify.aws/lib/datastore/data-access/q/platform/js#predicates) but not in AWS AppSync.
- AWS AppSync schemas do not follow the edges/nodes of the [Relay spec](https://relay.dev/docs/guides/graphql-server-specification) but are spiritually similar.
- You [cannot query by `page` number](https://github.com/aws-amplify/amplify-cli/issues/5086), you have to query by `nextToken`.

</Callout>

> ### Cancel read requests

Author: We should review where else this fragment is used and see if we want to change this and put it in the docs. It would mean we can make the paragraph more consistent and just replace code for other frameworks. This sample comes from https://github.com/aws-amplify/docs/blob/main/src/fragments/lib/graphqlapi/js/cancel-request.mdx

API.cancel(promise, "my error message");
```

> **Recap:** You now have an understanding of how to read your data through `Get` and `List` queries. 

Author: Anything to add to this recap as key take aways for real-world application?

Reviewer: maybe we should to a "final pass" on all the recaps to make sure the examples align towards the end of the JTBD project. I think then the content will also better line up if we use one real-world example/setup throughout.

## Subscribe to real-time events

> You can subscribe to real-time events to help manage your data. In the section below we outline the benefits of enabling real-time data integrations and how to set up and filter these subscriptions. We will also cover how to unsubscribe from subscriptions when needed. Before you begin you will need your application connected to the API and populated data to modify.

Author: In our workshop we discussed adding some context around the befits of enabling real-time data integrations. There isn't anything new here - is there more we wanted to add here or does this cover it?

</Block>
</BlockSwitcher>

</Accordion>

<br/>

> <Accordion title='Create a custom GraphQL subscription by ID' headingLevel='4' eyebrow='Walkthrough'>

Author: Added the other guide mentioned in the comments from before for this expander. This one reads more as an example use case so I wonder if we might move it to that page instead? Kept it so we can have the conversation on where it can go so we can begin getting rid of the guides menu items.

Reviewer: let's keep it in for now but I think there's opportunity to reduce this into a short blurb and move the main content to "custom business logic"

To implement this, you could update the schema with the following:

```graphql
type Post @model @auth(rules: [{ allow: public, provider: apiKey }]){
  id: ID!
  title: String!
  content: String
  comments: [Comment] @hasMany
}

type Comment @model @auth(rules: [{ allow: public, provider: apiKey }]){
  id: ID!
  content: String
  postCommentsId: ID!
}

type Subscription {
  onCommentByPostId(postCommentsId: ID!): Comment
    @aws_subscribe(mutations: ["createComment"])
}

```

Now you can create a custom subscription for comment creation with a specific post id:

```js
import { API } from 'aws-amplify';
import { onCommentByPostId } from './graphql/subscriptions';

API.graphql({
  query: onCommentByPostId,
  variables: {
    postCommentsId: "12345"
  }
})
.subscribe({
  next: data => {
    console.log('data: ', data)
  }
})
```

</Accordion>

### Unsubscribe from a subscription

You can also unsubscribe from events by using subscriptions by implementing the following:

```ts
// Stop receiving data updates from the subscription
sub.unsubscribe();
```

> **Recap:** You now have set up subscriptions for real-time events and understand how to filter and cancel these subscriptions when needed.

Author: Any tie-in to real-world use and application of the topic discussed OR reiterate the benefit to add here?

## Conclusion

Congratulations! You have finished the **Connect a data stack** guide. In this guide, you installed and configured Amplify Libraries; used GraphQL mutations for app data requests; filtered app data in Amplify; and set up, filtered, and canceled subscriptions for real-time events.

## Next steps

> Our recommended next steps include continuing to build out and customize your information architecture for your data. Some resources that will help with this work include:

Author: I don't see it on other pages but do we want to add 1-sentence descriptions on what these pages entail? They are also all CLI pages - any specific heads-up there?

Reviewer: IMHO, we could use Claude to summarize a one-sentence description. See if that lines up with our expectations. It'd be a good test to see if our content actually meets what we set them out to do as well? Otherwise I think the fastest way to pull one-sentence descriptions would be from the `const meta = { ... }` of those pages.
